pkg/pillar/pubsub/publish.go

```diff
@@ -130,149 +97,160 @@ func (pub *Publication) Unpublish(key string) error {
 	}
 	pub.updatersNotify(name)
 
-	fileName := pub.dirName + "/" + key + ".json"
-	log.Debugf("Unpublish deleting file %s\n", fileName)
-	if err := os.Remove(fileName); err != nil {
-		errStr := fmt.Sprintf("Unpublish(%s/%s): failed %s",
-			name, key, err)
-		return errors.New(errStr)
-	}
-	return nil
+	return pub.driver.Unpublish(key)
 }
 
-// SignalRestarted signal that a publication is restarted
+// SignalRestarted signal that this publication is restarted
 func (pub *Publication) SignalRestarted() error {
 	log.Debugf("pub.SignalRestarted(%s)\n", pub.nameString())
 	return pub.restartImpl(true)
 }
 
-// ClearRestarted clear the restart signal
+// ClearRestarted clear the restarted flag
 func (pub *Publication) ClearRestarted() error {
 	log.Debugf("pub.ClearRestarted(%s)\n", pub.nameString())
 	return pub.restartImpl(false)
 }
 
-// Record the restarted state and send over socket/file.
-func (pub *Publication) restartImpl(restarted bool) error {
-
-	name := pub.nameString()
-	log.Infof("pub.restartImpl(%s, %v)\n", name, restarted)
-
-	if restarted == pub.km.restarted {
-		log.Infof("pub.restartImpl(%s, %v) value unchanged\n",
-			name, restarted)
-		return nil
-	}
-	pub.km.restarted = restarted
-	if restarted {
-		// XXX lock on restarted to make sure it gets noticed?
-		// XXX bug?
-		// Implicit in updaters lock??
-		pub.updatersNotify(name)
-	}
-
-	restartFile := pub.dirName + "/" + "restarted"
-	if restarted {
-		f, err := os.OpenFile(restartFile, os.O_RDONLY|os.O_CREATE, 0600)
-		if err != nil {
-			errStr := fmt.Sprintf("pub.restartImpl(%s): openfile failed %s",
-				name, err)
-			return errors.New(errStr)
-		}
-		f.Close()
+// Get the value for a given key
+func (pub *Publication) Get(key string) (interface{}, error) {
+	m, ok := pub.km.key.Load(key)
+	if ok {
+		return m, nil
 	} else {
-		if err := os.Remove(restartFile); err != nil {
-			errStr := fmt.Sprintf("pub.restartImpl(%s): remove failed %s",
-				name, err)
-			return errors.New(errStr)
-		}
+		name := pub.nameString()
+		errStr := fmt.Sprintf("Get(%s) unknown key %s", name, key)
+		return nil, errors.New(errStr)
 	}
-	return nil
 }
 
-func (pub *Publication) serialize(sock net.Conn, keys []string,
-	sendToPeer localCollection) error {
+// GetAll enumerate all the key-value pairs for the collection
+func (pub *Publication) GetAll() map[string]interface{} {
```

Comment:
I am wondering if we should provide an iterate function along with getAll:

func (pub *Publication) Iterate(callbackFn) 

The prototype for CallbackFn would be:
   func IterateCallbackType(key string, value Interface{}) bool
     - Return Value true --> Continue iterating
     - Return Value false --> Stop iterating.

 90% of the cases we go GetAll is for iterating over them.. This function would be more useful..

Reply:
We can, but is that any easier than just doing:

```golang
for key, value := range pub.GetAll() {
  if doNotContinue {
     break
  }
}
```

go makes it pretty easy to iterate over collections. Either way, if we want it... separate PR.